app/code/Magento/Swatches/view/frontend/web/js/swatch-renderer.js

```diff
@@ -329,6 +329,28 @@ define([
         },
 
         /**
+         * Determine product id and related data
+         *
+         * @returns {{productId: *, isInProductView: boolean}}
+         * @private
+         */
+        _determineProductData: function() {
+            // Check if product is in a list of products.
+            var productId = this.element.parents('.product-item-details')
+                    .find('.price-box.price-final_price').attr('data-product-id'),
```

Comment:
Hello @zetlen ,

This sounds like a very good idea, but I feel it's out of the scope of this PR. There are other few things I noticed that need refactoring in swatch-renderer.js, but I tried to keep the code as similar as possible with the old code and just fix the issues, so the code review will be as easy as possible. I would rather leave the refactoring on Magento team's side.